### PR TITLE
Update distirbuted training APIs.

### PR DESCRIPTION
**Pull request by luomai**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Description

The number of epochs and data shuffling decisions are made by the users when they declare input training dataset.

### CODE REVIEW

File: tensorlayer/distributed.py [hunk 1]
```diff
@@ -218,17 +209,8 @@ def train_on_batch(self):
         """Train a mini-batch."""
         self._sess.run(self._train_op)
 
-    def train_to_end(self):
```
Comment (zsdonghao) [inline tensorlayer/distributed.py]: why train to end is removed?

Comment (JingqingZ) [reply tensorlayer/distributed.py]: `train_to_end` is actually `train_on_batch` with a while loop. This functionality can be easily implemented and also modified by users. In order to allow more flexibility for users, we may need to remove `train_to_end`.

File: examples/distributed_training/tutorial_cifar10_distributed_trainer.py [hunk 2]
```diff
@@ -22,10 +22,12 @@
 tl.logging.set_verbosity(tl.logging.DEBUG)
 
 
-def make_dataset(images, labels):
+def make_dataset(images, labels, num_epochs=1, shuffle_data_seed=0):
     img = tf.data.Dataset.from_tensor_slices(images)
     lab = tf.data.Dataset.from_tensor_slices(np.array(labels, dtype=np.int64))
-    return tf.data.Dataset.zip((img, lab))
+    dataset = tf.data.Dataset.zip((img, lab))
+    dataset = dataset.repeat(num_epochs).shuffle(buffer_size=10000, seed=shuffle_data_seed)
```
Comment (JingqingZ) [inline examples/distributed_training/tutorial_cifar10_distributed_trainer.py]: Why is the data firstly repeated and then shuffled? Just wondering if the `reshuffle_each_iteration` argument in [`tf.data.Dataset.shuffle()`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle) may achieve the functionality of shuffling each epoch?

Comment (luomai) [reply examples/distributed_training/tutorial_cifar10_distributed_trainer.py]: `dataset.repeat.shuffle` is equivalent to `reshuffle_each_iteration`; however, this operation is expensive when the dataset is big.

`dataset.shuffle.repeat` is relatively cheap.

File: tensorlayer/distributed.py (hunk 1 above)
Comment (luomai) [reply tensorlayer/distributed.py]: After talking to Jingqing yesterday, now I have a much more clear view about the trainer abstraction. In principle, we shall keep the trainer interface very flexible. For this purpose, we shall actually expose the internal session to the users and let them manipulate them. In addition we shall add a bunch of helper functions to help them validate the training error and adapt hyper-parameters based on epoch, instead of the global_steps which are very confused in a parallel setting. 

@zsdonghao @JingqingZ

**Comment by luomai** [comment]
@lgarithm could you advice how to re-trigger the [failed CI job](https://travis-ci.org/tensorlayer/tensorlayer/jobs/431863328)?
